notebooks/Mitigating-Disparities-in-Ranking-from-Binary-Data.ipynb

```diff
@@ -0,0 +1,592 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from fairlearn.widget import FairlearnDashboard\n",
+    "FairlearnDashboard(sensitive_features=A_test, sensitive_feature_names=['Race'],\n",
+    "                   y_true=y_test,\n",
+    "                   y_pred={\"unmitigated\": scores_unmitigated, \"model35\": scores_model35, \"model36\": scores_model36})"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Obtaining Low-Disparity Classifiers\n",
+    "\n",
+    "In this section, we shift attention from the task of scoring and ranking students to the task of automatically classifying students, for example, in order to screen them for an interview or a deeper review of their application materials. Our goal is to obtain a _classifier_ that maximizes AUC while respecting the demographic parity.\n",
+    "\n",
+    "The outputs of a classifier are either 0 or 1, so it is possible to re-interpret the AUC of a classifier as the *balanced accuracy*, meaning the accuracy under the distribution re-weighted to have the same mass of positive and negative examples. Demographic disparity can also be interpreted as the difference between the rates at which the students of either race are classified as 1; we refer to this rate as the _selection rate_.\n",
+    "\n",
+    "### Postprocessing\n",
+    "\n",
+    "We first show how to obtain low-disparity classifiers by thresholding scores&mdash;such as the scores produced by unmitigated logistic regression&mdash;using the postprocessing algorithm of [Hart et al. (2016)](https://arxiv.org/abs/1610.02413), which we call  `ThresholdOptimizer`. This algorithm finds thresholds that optimize accuracy subject to the constraint that there be no demographic disparity on the training data. Since our goal here is to optimize _balanced_ accuracy rather than accuracy, we first re-balance the data by randomly subsampling positive examples, so they are equal in number to negative examples."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from fairlearn.postprocessing import ThresholdOptimizer\n",
+    "\n",
+    "class LogisticRegressionAsRegression:\n",
+    "    def __init__(self, logistic_regression_estimator):\n",
+    "        self.logistic_regression_estimator = logistic_regression_estimator\n",
+    "    \n",
+    "    def fit(self, X, y):\n",
+    "        self.logistic_regression_estimator.fit(X, y)\n",
+    "    \n",
+    "    def predict(self, X):\n",
+    "        # use predict_proba to get real values instead of 0/1, select only prob for 1\n",
+    "        scores = self.logistic_regression_estimator.predict_proba(X)[:,1]\n",
+    "        return scores\n",
+    "\n",
+    "\n",
+    "balanced_index_pass0 = y_train[y_train==0].index \n",
+    "balanced_index_pass1 = y_train[y_train==1].sample(n=balanced_index_pass0.size, random_state=0).index\n",
```

Comment:
Should this be stratified, in order to get proportional representation from blacks and whites? Given the imbalance in numbers, it's entirely possible that `balanaced_index_pass1` could only have data from white students.

Reply:
stratification would change the metric optimized. the currently implemented sampling exactly corresponds to optimizing overall AUC (equal to balanced accuracy).

Reply:
Might want to comment on that. Again, it comes back to how much about fairness we assume people already know.

Reply:
I added a bit of a clarification--the rebalancing has nothing to do with fairness and it's only about the objective optimized. Also--I don't want to get bogged down by this too much, because we will be incorporating balanced accuracy into both post-processing and reductions, so this step won't be required down the line.